stl/inc/memory

```diff
@@ -3870,10 +3870,10 @@ protected:
     _Atomic_ptr_base(remove_extent_t<_Ty>* const _Px, _Ref_count_base* const _Ref) noexcept
         : _Ptr(_Px), _Repptr(_Ref) {}
 
-    void _Wait(remove_extent_t<_Ty>* _Old, memory_order) const noexcept {
+    void _Wait(remove_extent_t<_Ty>* _Old, _Ref_count_base* const _Old_Rep, memory_order) const noexcept {
         for (;;) {
             auto _Rep   = _Repptr._Lock_and_load();
-            bool _Equal = _Ptr.load(memory_order_relaxed) == _Old;
+            bool _Equal = _Ptr.load(memory_order_relaxed) == _Old && (_Old == nullptr || _Rep == _Old_Rep);

```

Comment:
I don't think this is a correct and complete fix.
It handles different control blocks in this check, but not in the `__std_atomic_wait_direct` below.

Reply:
Ok. If I understand correctly then we don't handle this situation:

1) `_Ptr == _Old` and `_Rep` == `_Old_Rep`
2) Call `__std_atomic_wait_direct(&_Ptr, &_Old);`
3) Change `_Old_Rep`, but `_Ptr == _Old` and we won't notice that  `_Rep` != `_Old_Rep` and still waiting...

Reply:
Actually I've looked into the implementation in more detail, and it is more complicated

We are waiting for `_Rep` to change, not `_Ptr` in `_Atomic_ptr_base`

So this does handle different control blocks, but does not seem to handle aliased shared pointers with one control blocks.

I think then your fix is a good thing, and regarding aliased pointers with single control blocks we can have a separate issue. This situation is hard to create anyway, most of the time aliased pointers have different types, so don't compare.

Reply:
Hm, no, we are waiting on `_Ptr`...